internal/agent/device/config/hook.go

```diff
@@ -0,0 +1,229 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/agent/client"
+	"github.com/flightctl/flightctl/pkg/executer"
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/fsnotify/fsnotify"
+)
+
+const (
+	DefaultHookActionTimeout = 10 * time.Second
+)
+
+var _ HookManager = (*hookManager)(nil)
+
+type hookManager struct {
+	mu            sync.Mutex
+	watcher       *fsnotify.Watcher
+	handlers      map[string]HookHandler
+	systemdClient *client.Systemd
+	exec          executer.Executer
+
+	log *log.PrefixLogger
+}
+
+// NewHookManager creates a new device action manager.
+func NewHookManager(log *log.PrefixLogger, exec executer.Executer) (HookManager, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	return &hookManager{
+		watcher:       watcher,
+		handlers:      make(map[string]HookHandler),
+		systemdClient: client.NewSystemd(exec),
+		exec:          exec,
+		log:           log,
+	}, nil
+}
+
+// Run starts the hook manager and listens for events.
+func (m *hookManager) Run(ctx context.Context) {
+	defer func() {
+		if err := m.watcher.Close(); err != nil {
+			m.log.Errorf("Error closing watcher: %v", err)
+		}
+		m.log.Infof("Hook manager stopped")
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event, ok := <-m.watcher.Events:
+			if !ok {
+				return
+			}
+			err := m.Handle(event)
+			if err != nil {
+				m.log.Errorf("error: %v", err)
+			}
+		case err, ok := <-m.watcher.Errors:
+			if !ok {
+				return
+			}
+			m.log.Errorf("error: %v", err)
+		}
+	}
+}
+
+// Update the manager with the new hook if appropriate.
+func (m *hookManager) Update(hook *v1alpha1.DeviceConfigHook) (bool, error) {
```

Comment:
add preflight check of all binary paths before we continue

Reply:
actually this should be on handle because the binaries could be added by config.